Prevent AI from queuing more than 1 MCV at a time

NotoriousPyro opened the pull request:
This change prevents the AI from spam-creating MCVs when it is below the MinimumConstructionYardCount. It disallows the AI from queuing up more than one new MCV at a time.

This change was due to the behaviour I noticed when playing with MinimumConstructionYardCount = 2, the AI would often build 2 or more MCVs at a time and I noticed the pattern was one queued per factory which could support it.

It effectively stops the AI from bankrupting itself by buying several MCVs.

NotoriousPyro committed:
Update McvManagerBotModule

Disallow the AI from queuing up more than 1 MCV at a time

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -132,14 +132,39 @@ void IBotTick.BotTick(IBot bot)
+
+			// Only build MCV if the conyardCount + mcvCount are lower than the minimum conyard count.
+			var conyardCount = AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player);
+			if (conyardCount + mcvCount >= Info.MinimumConstructionYardCount)
```
`mcvCount` can only ever be zero here (since a greater number would be filtered out above), so this can simplify to
`if (conyardCount >= Info.MinimumConstructionYardCount)`? (And update the comment.)

NotoriousPyro replied:
Good spot.

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -132,14 +132,39 @@ void IBotTick.BotTick(IBot bot)
+				return false;
+
+			// Only build MCV if there isn't already one being built.
+			var blds = player.World.ActorsWithTrait<ProductionQueue>()
```
Style nit: `blds` -> `queues`

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -132,14 +132,39 @@ void IBotTick.BotTick(IBot bot)
+				var q = factory.AllQueued().Where(z => Info.McvTypes.Contains(z.Item));
+
+				if (q != null)
+					queuedMcvCount += q.Count();
```
Seeing that we immediately return `false` in case there is a single mcv, this can simplify to something like:
```
for (var factory in queues)
	if (factory.AllQueued().Any(z => Info.McvTypes.Contains(z.Item)))
		return false;

NotoriousPyro committed:
Add rec. changes by OpenRA team

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -132,14 +132,31 @@ void IBotTick.BotTick(IBot bot)
+			if (mcvCount > 0)
+				return false;
+
+			// Only build MCV if the conyardCount + mcvCount are lower than the minimum conyard count.
```
This comment now wants the reference to `mcvCount` removed as well, I suppose.

Thinking about it, you should remove the first three comments (line 135, 140 and this one) as Imho your variables are well named so it is easy to understand what is going on already (i.e. the comments serve no purpose).

NotoriousPyro commented:
Whoops sorry, I am changing another commit and re-requested review on the wrong one ;)

NotoriousPyro committed:
Remove a few unneeded comments

NotoriousPyro committed:
Add delays to building MCVs

Add config option of how long the AI should wait in between building additional MCVs above the first

Add delay to prevent the AI queuing an additional MCV while it is still being delivered (nod)

NotoriousPyro commented:
I have added a change to delay how often the AI can queue up an MCV if one was queued previously. This was due to Nod air strips having a delay in delivery, the AI would count zero MCVs and queue another. The delay is 2000 ticks.

I also added a configuration for delaying the build of additional MCVs when there's already a conyard placed.

NotoriousPyro committed:
Move tick check for performance

pchote commented:
The issue from the description sounds like a flaw in the `unitBuilder.RequestedProductionCount(bot, mcvInfo.Name)` call. We really should aim to understand what is going wrong and fix that instead of implementing a completely separate method to query the production queue!

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -132,14 +139,38 @@ void IBotTick.BotTick(IBot bot)
+				.Select(a => a.Trait);
+
+			foreach (var factory in queues)
+				if (factory.AllQueued().Any(z => Info.McvTypes.Contains(z.Item)))
```
Style nit: The `foreach` should have braces since the inner block is more than one line.

NotoriousPyro replied:
You suggested it to be this way here: https://github.com/OpenRA/OpenRA/pull/17667#discussion_r377486968

Can you please make your mind up? :)

abcdefg30 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -132,14 +139,38 @@ void IBotTick.BotTick(IBot bot)
+			foreach (var factory in queues)
+				if (factory.AllQueued().Any(z => Info.McvTypes.Contains(z.Item)))
+				{
+					newMcvInterval = world.WorldTick + 2000;
```
Shouldn't this be unhardcoded too? Also why are we just resetting this timer together with `buildAdditionalMcvDelay`?

NotoriousPyro replied:
It's a performance tick to ensure that the AI doesn't try to request another MCV within the current WorldTick + 2000. The other parameter buildAdditionalMcvDelay is reset to WorldTick + configurable BuildAdditionalMcvDelay. It allows you to slow down the rate at which the AI is ultimately able to request new construction yards.

Mailaender replied:
Yes, but why not 
```suggestion
					newMcvInterval = world.WorldTick + Info.BuildMcvDelay;
```

pchote commented:
I've noted this PR in #19247 to track this for the future, but for now are closing this to help address the unsustainable state of the PR queue.
